federation/plugin/pkg/admission/schedulingpolicy/admission.go

```diff
@@ -0,0 +1,217 @@
+	maxRetryBackoff          = time.Second * 60
+	defaultRetryBackoff      = time.Millisecond * 100
+)
+
+type admissionConfig struct {
+	Kubeconfig   string        `json:"kubeconfig"`
+	RetryBackoff time.Duration `json:"retryBackoff"`
+}
+
+type admissionController struct {
+	*admission.Handler
+	policyEngineClient       *rest.RESTClient            // client to communicate with policy engine
+	policyEngineRetryBackoff time.Duration               // backoff for policy engine queries
+	client                   internalclientset.Interface // client to communicate with federation-apiserve
+}
+
+var _ = kubeapiserveradmission.WantsInternalKubeClientSet(&admissionController{})
+
+func newAdmissionController(file io.Reader) (*admissionController, error) {
+
+	config, err := loadConfig(file)
+	if err != nil {
+		return nil, err
+	}
+
+	policyEngineClient, err := loadRestClient(config.Kubeconfig)
+	if err != nil {
+		return nil, err
+	}
+
+	c := &admissionController{
+		Handler:                  admission.NewHandler(admission.Create, admission.Update),
+		policyEngineClient:       policyEngineClient,
+		policyEngineRetryBackoff: config.RetryBackoff,
+	}
+
+	return c, nil
+}
+
+func (c *admissionController) Validate() error {
+	if c.client == nil {
+		return fmt.Errorf("%s requires a client", pluginName)
+	}
+	return nil
+}
+
+func (c *admissionController) SetInternalKubeClientSet(client internalclientset.Interface) {
+	c.client = client
+}
+
+func (c *admissionController) Admit(a admission.Attributes) (err error) {
+
+	failClosed, err := c.shouldFailClosed()
+	if err != nil {
+		return c.handleError(a, true, err)
+	}
+
+	obj := a.GetObject()
+	decision, err := newPolicyEngineQuery(c.policyEngineClient, c.policyEngineRetryBackoff, a.GetUserInfo(), obj, a.GetKind()).Do()
+
+	if err != nil {
+		return c.handleError(a, failClosed, err)
+	}
+
+	if err := decision.Error(); err != nil {
+		return c.handleError(a, true, err)
+	}
+
+	mergeAnnotations(obj, decision.Resource.Metadata.Annotations)
+
+	return nil
+}
+
+func (c *admissionController) handleError(a admission.Attributes, failClosed bool, err error) error {
+
+	if !failClosed {
+		glog.V(4).Infof("an error occurred but request will be allowed (%v namespace does not contain any ConfigMap resources): %v.", policyConfigMapNamespace, err.Error())
+		return nil
+	}
+
+	c.publishEvent(a, err.Error())
+
+	return admission.NewForbidden(a, err)
+}
+
+func (c *admissionController) publishEvent(a admission.Attributes, msg string) {
+
+	obj := a.GetObject()
+
+	ref, err := ref.GetReference(api.Scheme, obj)
+	if err != nil {
+		glog.Errorf("get reference for %v (kind: %v) failed: %v", a.GetName(), a.GetKind().Kind, err.Error())
+		return
+	}
+
+	event := &api.Event{
+		InvolvedObject: *ref,
+		Message:        msg,
+		Source: api.EventSource{
+			Component: fmt.Sprintf("schedulingpolicy"),
+		},
+		Type: "Warning",
+	}
+
+	if _, err := c.client.Core().Events(a.GetNamespace()).Create(event); err != nil {
```

Comment:
It seems the EventRecorder utility is not compatible with the internalclientset. Let me know how we should proceed.